File: core/src/main/java/io/github/shaksternano/borgar/core/media/MediaReaders.java
package io.github.shaksternano.borgar.core.media;

import io.github.shaksternano.borgar.core.exception.UnreadableFileException;
import io.github.shaksternano.borgar.core.media.readerold.*;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Map;

public class MediaReaders {

    private static final Map<String, MediaReaderFactoryOld<ImageFrameOld>> imageReaderFactories = new HashMap<>();
    private static final Map<String, MediaReaderFactoryOld<AudioFrameOld>> audioReaderFactories = new HashMap<>();

    public static MediaReader<ImageFrameOld> createImageReader(File media, String format) throws UnreadableFileException {
        var factory = getImageReaderFactory(format);
        return createReader(factory, media, format);
    }

    public static MediaReader<ImageFrameOld> createImageReader(InputStream media, String format) throws UnreadableFileException {
        var factory = getImageReaderFactory(format);
        return createReader(factory, media, format);
    }

    private static MediaReaderFactoryOld<ImageFrameOld> getImageReaderFactory(String format) {
        return imageReaderFactories.getOrDefault(format.toLowerCase(), FFmpegImageReader.Factory.INSTANCE);
    }

    public static MediaReader<AudioFrameOld> createAudioReader(File media, String format) throws UnreadableFileException {
        var factory = getAudioReaderFactory(format);
        return createReader(factory, media, format);
    }

    @SuppressWarnings("unused")
    public static MediaReader<AudioFrameOld> createAudioReader(InputStream media, String format) throws UnreadableFileException {
        var factory = getAudioReaderFactory(format);
        return createReader(factory, media, format);
    }

    private static MediaReaderFactoryOld<AudioFrameOld> getAudioReaderFactory(String format) {
        return audioReaderFactories.getOrDefault(format.toLowerCase(), FFmpegAudioReader.Factory.INSTANCE);
    }

    private static <T> MediaReader<T> createReader(MediaReaderFactoryOld<T> factory, File media, String format) throws UnreadableFileException {
        try {
            return factory.createReader(media, format);
        } catch (IOException e) {
            throw new UnreadableFileException(e);
        }
    }

    private static <T> MediaReader<T> createReader(MediaReaderFactoryOld<T> factory, InputStream media, String format) throws UnreadableFileException {
        try {
            return factory.createReader(media, format);
        } catch (IOException e) {
            throw new UnreadableFileException(e);
        }
    }

    private static void registerImageReaderFactory(MediaReaderFactoryOld<ImageFrameOld> factory, String... formats) {
        for (var format : formats) {
            imageReaderFactories.putIfAbsent(format.toLowerCase(), factory);
        }
    }

    @SuppressWarnings("SameParameterValue")
    private static void registerAudioReaderFactory(MediaReaderFactoryOld<AudioFrameOld> factory, String... formats) {
        for (var format : formats) {
            audioReaderFactories.putIfAbsent(format.toLowerCase(), factory);
        }
    }

    private static void registerImageOnlyReaderFactory(MediaReaderFactoryOld<ImageFrameOld> factory, String... formats) {
        registerImageReaderFactory(factory, formats);
        registerAudioReaderFactory(NoAudioReader.Factory.INSTANCE, formats);
    }

    static {
        registerImageOnlyReaderFactory(ScrimageGifReader.Factory.INSTANCE,
            "gif"
        );
        registerImageOnlyReaderFactory(JavaxImageReader.Factory.INSTANCE,
            "bmp",
            "jpeg",
            "jpg",
            "wbmp",
            "png",
            "gif",
            "tif",
            "tiff"
        );
        registerImageOnlyReaderFactory(WebPImageReader.Factory.INSTANCE,
            "webp"
        );
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/MediaReaders.kt
package io.github.shaksternano.borgar.core.media

import io.github.shaksternano.borgar.core.collect.putAllKeys
import io.github.shaksternano.borgar.core.exception.UnreadableFileException
import io.github.shaksternano.borgar.core.io.DataSource
import io.github.shaksternano.borgar.core.io.fileFormat
import io.github.shaksternano.borgar.core.media.reader.*

private val imageReaderFactories: MutableMap<String, ImageReaderFactory> = mutableMapOf()

private val audioReaderFactories: MutableMap<String, AudioReaderFactory> = mutableMapOf()

@Suppress("unused")
private val init: Unit = run {
    registerImageOnlyFactory(JavaxImageReader.Factory)
    registerImageOnlyFactory(GifReader.Factory)
    registerImageOnlyFactory(WebPImageReader.Factory)
    registerImageOnlyFactory(PdfReader.Factory)
}

private fun registerImageFactory(factory: ImageReaderFactory) =
    imageReaderFactories.putAllKeys(
        factory.supportedFormats,
        factory,
    )

private fun registerAudioFactory(factory: AudioReaderFactory) =
    audioReaderFactories.putAllKeys(
        factory.supportedFormats,
        factory,
    )

private fun registerImageOnlyFactory(factory: ImageReaderFactory) {
    registerImageFactory(factory)
    registerAudioFactory(NoAudioReader.Factory(factory.supportedFormats))
}

fun isReaderFormatSupported(format: String): Boolean =
    imageReaderFactories.containsKey(format)

suspend fun createImageReader(input: DataSource): ImageReader =
    createImageReader(input, input.fileFormat())

suspend fun createImageReader(input: DataSource, format: String): ImageReader {
    val factory = imageReaderFactories.getOrDefault(format, FFmpegImageReader.Factory)
    return try {
        factory.create(input)
    } catch (t: Throwable) {
        throw UnreadableFileException(t)
    }
}

suspend fun createAudioReader(input: DataSource): AudioReader =
    createAudioReader(input, input.fileFormat())

suspend fun createAudioReader(input: DataSource, format: String): AudioReader {
    val factory = audioReaderFactories.getOrDefault(format, FFmpegAudioReader.Factory)
    return try {
        factory.create(input)
    } catch (t: Throwable) {
        throw UnreadableFileException(t)
    }
}
